website/statistics.py

```diff
@@ -290,20 +279,651 @@ def get_grid_info(self, user, class_id, level):
+            adventures=adventures,
+            attempted_adventures=attempted_adventures,
+            max_level=HEDY_MAX_LEVEL,
+            selected_item=selected_item,
+            current_page='my-profile'
+        )
+
+    @route("/live_stats/class/<class_id>/error/<error_id>", methods=["DELETE"])
+    @requires_login
+    def remove_common_error_item(self, user, class_id, error_id):
+        """
+        Removes the common error item by setting the active flag to 0.
+        """
+        common_errors = self.__common_errors(class_id)
+        for i in range(len(common_errors['errors'])):
+            if common_errors['errors'][i]['id'] == int(error_id) and common_errors['errors'][i]['active'] == 1:
+                common_errors['errors'][i]['active'] = 0
+                self.db.update_class_errors(common_errors)
+                break
+
+        return {}, 200
+
+    def retrieve_exceptions_per_student(self, class_id):
+        """
+        Retrieves exceptions per student in the class
+        :param class_id: class id
+        :return: exceptions_per_user
+        """
+        class_ = self.db.get_class(class_id)
+        exceptions_per_user = {}
+        students = sorted(class_.get("students", []))
+        for student_username in students:
+            program_stats = self.db.get_program_stats([student_username], None, None)
+            if program_stats:
+                # if there are multiple weeks, only get the most recent week's data
+                program_stats = program_stats[-1]
+                exceptions = {k: v for k, v in program_stats.items() if k.lower().endswith("exception")}
+                exceptions_per_user[student_username] = exceptions
+
+        return exceptions_per_user
+
+    def new_id_calc(self, common_errors, class_id):
+        """
+        Calculates the new id for a new common error entry.
+        :param common_errors: common errors from db
+        :param class_id: class id
+        :return: new id
+        """
+        common_error_ids = [int(x['id']) for x in common_errors['errors']]
+        new_id = max(common_error_ids) + 1 if common_error_ids else 0
+
+        # reached max common errors
+        if new_id > 0 and new_id % self.MAX_COMMON_ERRORS == 0:
+            # find all disables entries
+            disables = [x['id'] for x in common_errors['errors'] if x['active'] == 0]
+            if disables:
+                # assign oldest not used id to new error
+                new_id = disables[0]
+            else:
+                # forcefully overwrite oldest error despite not being resolved and set oldest half of the db to
+                # inactive to free up space
+                # Todo: could use a better way to handle this
+                new_id = 0
+
+                for i in range(self.MAX_COMMON_ERRORS // 2):
+                    common_errors['errors'][i]['active'] = 0
+                self.db.update_class_errors(common_errors)
+
+        return new_id
```

Comment:
I don't know if this is really necessary. This enforces a limit of 10 exception groups per class, so if the students of a class have made exceptions belonging to 10 different groups, it'll delete the 5 previous ones to make room for new. I think this is for not to clutter the UI. What do you think, @Felienne, do we leave this?